### PR TITLE
feat: Fuzzy Join support in EVA

### PR DESCRIPTION
**Pull request by ashmitaraju**
This PR

- [x] Introduces support for Fuzzy joins in EVA
- [x] It is implemented as a UDF but uses a nested loop join executor that can be generalized to nested loop joins in the future
- [x] Covers binder, optimizer and executor
- [x] Uses Levenshtein distance to calculate distance

### CODE REVIEW

File: eva/executor/fuzzy_join_executor.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+from typing import Generator, Iterator
+
+from eva.executor.abstract_executor import AbstractExecutor
+from eva.executor.executor_utils import ExecutorError, apply_predicate
+from eva.models.storage.batch import Batch
+from eva.plan_nodes.fuzzy_join_plan import FuzzyNestedJoinPlan
+from eva.utils.logging_manager import logger
+
+
+class FuzzyJoinExecutor(AbstractExecutor):
```
Comment (gaurav274) [inline eva/executor/fuzzy_join_executor.py:24]: Maybe rename it to `NestedLoopJoin`

File: eva/udfs/ndarray/fuzzy_join.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import pandas as pd
+
+from thefuzz import fuzz
+
+from eva.udfs.abstract.abstract_udf import AbstractUDF
+
+class FuzzyJoin(AbstractUDF):
```
Comment (gaurav274) [inline eva/udfs/ndarray/fuzzy_join.py:21]: Rename it to FuzzDistance or something similar

File: test/integration_tests/test_fuzzy_join.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+                    IMPL "{}/udfs/{}/fuzzy_join.py";
+        """.format(
+    EVA_INSTALLATION_DIR, "ndarray"
+)
+        execute_query_fetch_all(fuzzy_udf)
+    
+        fuzzy_join_query = """SELECT * FROM MyVideo a JOIN MyVideoCSV b
+                      ON FuzzyJoin(a.id, b.id) = 100;"""
+
+        actual_batch = execute_query_fetch_all(fuzzy_join_query)
```
Comment (gaurav274) [inline test/integration_tests/test_fuzzy_join.py:86]: Please also add an example of practical usecase

**Review by gaurav274 — CHANGES_REQUESTED** [review]

File: eva/executor/fuzzy_join_executor.py (hunk 1 above)
Comment (kaushikravichandran) [reply eva/executor/fuzzy_join_executor.py:24]: Could have a base `NestedLoopJoin` class, and FuzzyJoin could be a special type of it

File: eva/udfs/ndarray/fuzzy_join.py (hunk 2 above)
Comment (ashmitaraju) [reply eva/udfs/ndarray/fuzzy_join.py:21]: Done

File: test/integration_tests/test_fuzzy_join.py (hunk 3 above)
Comment (ashmitaraju) [reply test/integration_tests/test_fuzzy_join.py:83]: Will work on that. I think that fits better in a separate notebook, right?

File: eva/optimizer/rules/rules.py [hunk 4]
```diff
@@ -862,7 +860,10 @@ def promise(self):
         return Promise.LOGICAL_JOIN_TO_PHYSICAL_HASH_JOIN
 
     def check(self, before: Operator, context: OptimizerContext):
-        return before.join_type == JoinType.INNER_JOIN
+        j_child: FunctionExpression = before.join_predicate.children[0]
+        return before.join_type == JoinType.INNER_JOIN and (
```
Comment (gaurav274) [inline eva/optimizer/rules/rules.py:864]: Please add a comment here describing what is special about FuzzyDistance

File: eva/optimizer/rules/rules.py [hunk 5]
```diff
@@ -893,6 +894,31 @@ def apply(self, join_node: LogicalJoin, context: OptimizerContext):
...
+        pattern.append_child(Pattern(OperatorType.DUMMY))
+        pattern.append_child(Pattern(OperatorType.DUMMY))
+        super().__init__(RuleType.LOGICAL_JOIN_TO_PHYSICAL_NESTED_LOOP_JOIN, pattern)
+
+    def promise(self):
+        return Promise.LOGICAL_JOIN_TO_PHYSICAL_NESTED_LOOP_JOIN
+
+    def check(self, before: LogicalJoin, context: OptimizerContext):
+        j_child: FunctionExpression = before.join_predicate.children[0]
+        return before.join_type == JoinType.INNER_JOIN and j_child.name.startswith(
```
Comment (gaurav274) [inline eva/optimizer/rules/rules.py:909]: As comment as above

**Review by gaurav274 — APPROVED** [review]
Add comments and we are good to go

**Comment by gaurav274** [comment]
You need to check the failing test cases before merging. Thanks!

**Pull request merged**